web/modules/custom/joinup_group/src/Entity/PinnableGroupContentTrait.php

```diff
@@ -0,0 +1,113 @@
+
+use Drupal\Core\Field\Plugin\Field\FieldType\EntityReferenceItem;
+use Drupal\Core\TypedData\Exception\MissingDataException;
+
+/**
+ * Reusable methods for pinnable group content entities.
+ *
+ * @todo Once we are on PHP 7.3 the JoinupBundleClassMetaEntityTrait
+ *   should be included here.
+ */
+trait PinnableGroupContentTrait {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isPinned(?GroupInterface $group = NULL): bool {
+    if (!$meta_entity = $this->getMetaEntity('pinned_in')) {
+      return FALSE;
+    }
+
+    /** @var \Drupal\joinup_federation\RdfEntityReferenceFieldItemList $item_list */
+    $item_list = $meta_entity->get('field_pinned_in');
+
+    if (empty($group)) {
+      return !$item_list->isEmpty();
+    }
+
+    return array_reduce($item_list->getValue(), function (bool $carry, array $properties) use ($group) {
+      return $carry || ($properties['target_id'] ?? '') === $group->id();
+    }, FALSE);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function pin(?GroupInterface $group = NULL): PinnableGroupContentInterface {
+    // Default to the parent group.
+    $group = $group ?: $this->getGroup();
+
+    // Only update the pinned status in the database if needed.
+    if (!$this->isPinned($group)) {
+      $meta_entity = $this->getMetaEntity('pinned_in');
+      if (empty($meta_entity)) {
+        $meta_entity = $this->createMetaEntity('pinned_in');
+      }
+      $meta_entity->get('field_pinned_in')->appendItem($group->id());
+      $meta_entity->save();
+    }
+
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function unpin(?GroupInterface $group = NULL): PinnableGroupContentInterface {
+    // Default to the parent group.
+    $group = $group ?: $this->getGroup();
+
+    // Only update the pinned status in the database if needed.
+    if ($this->isPinned($group)) {
+      $meta_entity = $this->getMetaEntity('pinned_in');
+
+      /** @var \Drupal\joinup_federation\RdfEntityReferenceFieldItemList $item_list */
+      $item_list = $meta_entity->get('field_pinned_in');
+
+      $item_list->filter(function (EntityReferenceItem $item) use ($group) {
+        return $item->target_id !== $group->id();
+      });
+
+      // If the entity is no longer pinned in any groups, delete the meta entity
+      // rather than updating it. This ensures that the database will not retain
+      // a large number of empty meta entities for old content that was pinned
+      // at some point in the past but will probably never be pinned again.
+      if ($item_list->isEmpty()) {
+        $meta_entity->delete();
+      }
+      else {
+        $meta_entity->save();
+      }
+    }
+
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getPinnedGroupIds(): array {
+    $ids = [];
+
+    $meta_entity = $this->getMetaEntity('pinned_in');
+    foreach ($meta_entity->get('field_pinned_in') as $item) {
+      if ($item instanceof EntityReferenceItem) {
```

Comment:
It can be `NULL`. This `instanceof` is a shortcut that checks two things at once:

1. It asserts that the `$item` is not `NULL`.
2. Drupal core is not using strict typing, so to make this code work reliably we need to verify the type at this point.

Reply:
Honesty, I haven't seen until now such a check for a field that WE KNOW it's an entity reference. Neither can I imagine how this could be `NULL`. Looking to the `::unpin()` method, I see that it takes all precautions in this regard. Regarding strict typing, that is only an IDE issue but we can simply add the type hint:
```php
/** @var \Drupal\Core\Field\Plugin\Field\FieldType\EntityReferenceItem $item */
```